include/pybind11/numpy.h

```diff
@@ -120,6 +120,25 @@ inline numpy_internals &get_numpy_internals() {
     return *ptr;
 }
 
+PYBIND11_NOINLINE module_ import_numpy_core_submodule(const char *submodule_name) {
+    try {
+        return module_::import((std::string("numpy._core.") + submodule_name).c_str());
+    } catch (error_already_set &ex) {
+        if (!ex.matches(PyExc_ImportError)) {
+            throw;
+        }
+        try {
+            return module_::import((std::string("numpy.core.") + submodule_name).c_str());
+        } catch (error_already_set &ex) {
+            if (!ex.matches(PyExc_ImportError)) {
+                throw;
+            }
+            throw import_error(std::string("pybind11 couldn't import ") + submodule_name
+                               + " from numpy.");
+        }
+    }
+}
```

Comment:
What you have here looks acceptable, but I wonder if we could make this more robust pretty easily, along the lines of:

* `import numpy`
* Inspect `numpy.__version__` or similar to decide if `core` or `_core` is needed, and try only that one import. (I'm hoping this might be more feasible than usual because you own the numpy API.)

This is less likely to be surprising if there is some low-level issue in pybind11 or Python itself (refactoring, new major dev version), or numpy is changed further in the future. The error message then could be the original import error, re-raised using `raise_from` (very easy):

https://github.com/pybind/pybind11/blob/7e5edbc947168fa73e25c5b6e58d656f5e2e9863/include/pybind11/pytypes.h#L817

Reply:
So I did similar changes for JAX and originally I also decided to go with numpy version checks in the `if` statement and then a single import.  
But I was advised that `try ... catch` approach is more robust. Here's an explanation from Jake: https://github.com/google/jax/pull/16972#pullrequestreview-1563181273

>Version checks are fine for tests, but I'd prefer to avoid them in package import paths – e.g. imagine a user who has some custom numpy build with a version string that we don't anticipate, and then import jax might fail on trying to cast the version to a tuple of ints.

But I'm happy to switch to a version check if you prefer it!

Reply:
We're dealing with 1 try vs 2 tries here, that's similar but not the same.

> version check

I'm thinking of it more as a "capability check".

(Version numbers a convenient way to answer "what exactly do I have", that's a different purpose.)

I'm still hoping that we can rise to a level of organization that allows us to only have 1 import and **report the original error**: does or can numpy provide a simple API that allows us to only try once?

Reply:
Sure! I switched to an `if` statement that selects package path based on numpy version - now `(_)core.multiarray` or `(_)core._internal` is imported only once.  
(it might look lengthy, but this is recommended way to get major version integer: https://numpy.org/doc/stable/release/1.9.0-notes.html#numpyversion-class-added)

Reply:
(I think I somehow lost a comment here that I thought I typed in already. Sorry if something appears twice.)

```
std::string numpy_core_path = major_version >= 2 ? "numpy._core" : "numpy.core";
```

What you have looks fine though, if you prefer, the only thing I'd do is remove the explicit `std::string()` inside the two `if` branches, i.e. I think simply `numpy_core_path = "numpy._core";` will work and is what people usually do.

Reply:
I can’t really check (at a conference), but what are we using the core module for? If it’s private that means NumPy is free to break us any time. Would be very nice to start moving to whatever is supported instead longer term.

Reply:
Is it only our tests?

Reply:
@henryiii It's private and yes NumPy can break us at any time. @rwgk and I discuss a long term solution in this thread, which is moving us to Numpy's public Python API instead of their private C API.

Reply:
(Numpy has a public C API as well, but that's difficult for us to use because that would complicate our build process quite a bit).